superset/models/core.py

```diff
@@ -368,14 +372,38 @@ def get_sqla_engine_with_context(
         schema: Optional[str] = None,
         nullpool: bool = True,
         source: Optional[utils.QuerySource] = None,
+        override_ssh_tunnel: Optional["SSHTunnel"] = None,
     ) -> Engine:
-        yield self._get_sqla_engine(schema=schema, nullpool=nullpool, source=source)
+        ssh_params: Dict[str, Any] = {}
+        from superset.databases.dao import (  # pylint: disable=import-outside-toplevel
+            DatabaseDAO,
+        )
+
+        if ssh_tunnel := override_ssh_tunnel or DatabaseDAO.get_ssh_tunnel(
+            database_id=self.id
+        ):
+            # if ssh_tunnel is available build engine with information
+            url = make_url_safe(self.sqlalchemy_uri_decrypted)
+            ssh_tunnel.bind_host = url.host
+            ssh_tunnel.bind_port = url.port
+            ssh_params = ssh_tunnel.parameters()
+            with sshtunnel.open_tunnel(**ssh_params) as server:
+                yield self._get_sqla_engine(
+                    schema=schema,
+                    nullpool=nullpool,
+                    source=source,
+                    ssh_tunnel_server=server,
+                )
+
+        else:
+            yield self._get_sqla_engine(schema=schema, nullpool=nullpool, source=source)
```

Comment:
You can write this a bit more cleanly:

```python
if ssh_tunnel := ...:
    ...
    engine_context = sshtunnel.open_tunnel(**ssh_params)
else:
    engine_context = contextlib.nullcontext()

with engine_context as context:
    yield self._get_sqla_engine(schema=schema, nullpool=nullpool, source=source, ssh_tunnel_server=context)
```

Also, ideally we wouldn't have to pass `ssh_tunnel_server=server` to `_get_sqla_engine`, since it would be nice to decouple the SSH tunnel from it. Maybe passing the SQL Alchemy URI directly here.